### PR TITLE
kvserver: remove checksum computation tasks GC

### PR DESCRIPTION
**Pull request by pav-kv**
GC is no longer needed because the task is synchronized with the collection request, and both wait only up to 5s before giving up. Both are async, and normally the task runs O(tens of second) anyway, so it's ok if the failing path takes 5s.

This change makes the `replicaChecksum` lifecycle code significantly simpler.

Touches #77432

Release note: None

### CODE REVIEW

File: pkg/kv/kvserver/replica_consistency_test.go [hunk 1]
```diff
@@ -141,20 +138,18 @@ func TestGetChecksumNotSuccessfulExitConditions(t *testing.T) {
 	rc, err = tc.repl.getChecksum(ctx, id)
 	require.ErrorIs(t, err, context.Canceled)
 	require.Nil(t, rc.Checksum)
-	require.ErrorContains(t, startChecksumTask(context.Background(), id),
-		"checksum collection request gave up")
 
 	// The task failed to start because the checksum collection request did not
 	// join. Later, when it joins, it finds out that the task gave up.
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_consistency_test.go:143]: The comment needs update.

File: pkg/kv/kvserver/replica_consistency.go [hunk 2]
```diff
@@ -406,21 +386,11 @@ func (r *Replica) runConsistencyCheck(
...
-		}
-	}
-}
-
-// getReplicaChecksum returns replicaChecksum tracker for the given ID, and
-// whether it is still active (i.e. has a zero GC timestamp).
-func (r *Replica) getReplicaChecksum(id uuid.UUID, now time.Time) (*replicaChecksum, bool) {
+// getReplicaChecksum returns replicaChecksum tracker for the given ID. The
+// caller must call delReplicaChecksum when finished working on this tracker.
+func (r *Replica) getReplicaChecksum(id uuid.UUID) *replicaChecksum {
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/replica_consistency.go:391]: How about returning a cleanup closure instead, to make this requirement more explicit? E.g.:

```go
c, cleanup := r.getReplicaChecksum()
defer cleanup()
```

We could also consider renaming to something like `trackReplicaChecksum` to make it clearer that this isn't a simple read operation. No strong opinion either way, your call.

**Review by erikgrinaker — APPROVED** [review]
Lovely, thanks!

File: pkg/kv/kvserver/replica_consistency.go (hunk 2 above)
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_consistency.go:391]: Nice idea, makes the contract self-explanatory and removes the possibility of passing in a wrong pair of `(*replicaChecksum, id)`. Done.

File: pkg/kv/kvserver/replica_consistency_test.go (hunk 1 above)
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_consistency_test.go:143]: Done.

**Comment by pavelkalinnikov** [comment]
bors r=erikgrinaker

**Pull request merged**